### PR TITLE
Fix way of copying TestNGOptions

### PR DESCRIPTION
**Pull request by snoopcheri**
When copying the `TestNGTestFramework` in the test-retry Gradle plugin, this currently fails miserably with:

```
Caused by: java.lang.IllegalStateException: Cannot create DslObject for 'org.gradle.api.tasks.testing.testng.TestNGOptions@5b92dc57' (class: TestNGOptions) as it does not implement 'org.gradle.api.internal.IConventionAware' (it is not a DSL object)
	at org.gradle.api.internal.plugins.DslObject.toType(DslObject.java:108)
	at org.gradle.api.internal.plugins.DslObject.getConventionMapping(DslObject.java:83)
```

(build scan for test-retry Gradle plugin [here](https://ge.gradle.org/s/3f4cnm55m6oq2/tests/overview?outcome=failed))

It seems that the copied `TestNGOption` object must be created via an object factory.

This PR fixes this issue.

### CODE REVIEW

**Review by jvandort — CHANGES_REQUESTED** [review]
We should make this same change for `JUnitTestFramework` and `JUnitPlatformTestFramework`.

**Comment by snoopcheri** [comment]
> We should make this same change for `JUnitTestFramework` and `JUnitPlatformTestFramework`.

Do you mean for consistency, the `JUnitOptions` and `JUnitPlatformOptions` should also just provide a copy-method instead of a "copy-constructor"?

The `TestNGOptions` is the only one that has to be created via an object factory. The other two were already created with the default constructor before (see [here](https://github.com/gradle/gradle/blob/release/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestFramework.java#L47) and [here](https://github.com/gradle/gradle/blob/release/subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junitplatform/JUnitPlatformTestFramework.java#L50))

I figured that correctness-by-construction would be more important here, but I can surely make the copying consistent across all three versions.

File: subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/testng/TestNGOptions.java [hunk 1]
```diff
@@ -88,10 +88,10 @@ public TestNGOptions(ProjectLayout projectLayout) {
     }
 
     /**
-     * Copies the TestNG options.
+     * Copies the options from the source options into the current one.
      * @since 8.0
      */
-    public TestNGOptions(TestNGOptions other) {
+    public void copyFrom(TestNGOptions other) {
```
Comment (snoopcheri) [inline subprojects/testing-jvm/src/main/java/org/gradle/api/tasks/testing/testng/TestNGOptions.java:94]: The `addAll` calls are not really correct as we're not doing this in the constructor anymore. We should clear the collections first before populating them again.

**Comment by jvandort** [comment]
@snoopcheri My understanding is that all DSL instances that users interact with should be instantiated with via the instantiator. Obviously, my understanding is flawed 'cause this has worked in the past without us needing to instantiate `JUnitOptions ` and `JUnitPlatformOptions`. So perhaps we do not need to instantiate these with the instantiator. I will look into this myself and we don't need to block this PR on it. 

However, I would still prefer we copy all of the options files in the same manner. Given that Test NG requires a copy method and not a copy constructor, I would prefer we change the other options to use a copy method. 

Another option, which would require less changes (and TBH is probably preferable here) is to use the instantiator with the copy constructor. The instantiator will allow you to pass arguments into the constructor. Consider the following code:

```
TestNGOptions copiedOptions = objects.newInstance(TestNGOptions.class, options);
```
This way we can still use the copy constructor the way we have been.

**Comment by snoopcheri** [comment]
@jvandort 

I've tried to create `TestNGOptions` with the object factory, but that failed:
* if the constructor is not annotated with `@Inject`, it doesn't get picked up
* if there are multiple constructors annotated with `@Inject`, this fails with "Class TestNGOptions has multiple constructors that are annotated with @Inject."

Therefore I propose the following:
* remove all copy-constructors
* add copy-methods for all test options and make sure that containers are cleared before adding via `addAll(...)`

WDYT?

**Comment by jvandort** [comment]
@snoopcheri I would support your proposed change

**Comment by snoopcheri** [comment]
@jvandort I've implemented to copy methods

And I've added the following changes:
* copy-methods now have a simple unit test
* I stopped copying two fields from `TestNGOptions` as those are marked as transient. I don't believe we should copy those

**Comment by jvandort** [comment]
Looks like you've got an unused import in `JUnitPlatformOptions`. Try running `./gradlew sanityCheck` locally to make sure there aren't any more code quality issues after that

**Review by jvandort — APPROVED** [review]
LGTM other than sanity check failing

**Review by marcphilipp — APPROVED** [review]

**Comment by snoopcheri** [comment]
> LGTM other than sanity check failing

@jvandort All checks are passing now

**Comment by jvandort** [comment]
@bot-gradle test and merge

**Pull request merged**